src/test/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBulkRecordReaderTest.java

```diff
@@ -313,6 +317,153 @@ private void assertRecordReaderOutputRecords(String[] csvStrings, Schema schema,
     }
   }
 
+  /**
+   * The FailsafeException is thrown with the expected error message.
+   * The retry policy is correctly configured with the specified configurations.
+   */
+  @Test(expected = SalesforceQueryExecutionException.class)
+  public void testRetryMechanism() throws Exception {
+    Long initialRetryDuration = 5L;
+    Long maxRetryDuration = 80L;
+    Integer maxRetryCount = 5;
+    String csvString1 = "\"Id\",\"IsDeleted\",\"ExpectedRevenue\",\"LastModifiedDate\",\"CloseDate\",\"Time\"\n" +
+      "\"0061i000003XNcBAAW\",\"false\",\"1500.0\",\"2019-02-22T07:03:21.000Z\",\"2019-01-01\",\"12:00:30.000Z\"\n";
+    String csvString2 = "\"Id\",\"IsDeleted\",\"ExpectedRevenue\",\"LastModifiedDate\",\"CloseDate\",\"Time\"\n" +
+      "\"0061i000003XNcCAAW\",\"false\",\"112500.0\",\"2019-02-22T07:03:21.000Z\",\"2018-12-20\",\"12:00:40.000Z\"\n" +
+      "\"0061i000003XNcDAAW\",\"false\",\"220000.0\",\"2019-02-22T07:03:21.000Z\",\"2018-11-15\",\"12:00:50.000Z\"\n";
+
+    Schema schema = Schema.recordOf("output",
+      Schema.Field.of("Id", Schema.of(Schema.Type.STRING)),
+      Schema.Field.of("IsDeleted", Schema.of(Schema.Type.BOOLEAN)),
+      Schema.Field.of("ExpectedRevenue", Schema.of(Schema.Type.DOUBLE)),
+      Schema.Field.of("LastModifiedDate", Schema.of(Schema.LogicalType.TIMESTAMP_MICROS)),
+      Schema.Field.of("CloseDate", Schema.of(Schema.LogicalType.DATE)),
+      Schema.Field.of("Time", Schema.of(Schema.LogicalType.TIME_MICROS))
+    );
+
+    Assert.assertEquals(5, SalesforceSplitUtil.getRetryPolicy
+      (initialRetryDuration, maxRetryDuration, maxRetryCount).getConfig().getMaxRetries());
+      assertRecordReaderOutputRecordsRetryMechanism(new String[]{csvString1, csvString2}, schema);
+  }
+
+  private void assertRecordReaderOutputRecordsRetryMechanism(String[] csvStrings, Schema schema) throws Exception {
+    String jobId = "job";
+    String batchId = "batch";
+    String[] resultIds = new String[csvStrings.length];
+    for (int i = 0; i < csvStrings.length; i++) {
+      resultIds[i] = String.format("result%d", i);
+    }
+
+    SalesforceBulkRecordReader reader = new SalesforceBulkRecordReader(schema, jobId, batchId, resultIds);
+    BulkConnection mock = Mockito.mock(BulkConnection.class);
+    FieldSetter.setField(reader, SalesforceBulkRecordReader.class.getDeclaredField("bulkConnection"), mock);
+    for (int i = 0; i < csvStrings.length; i++) {
+      AsyncApiException salesforceQueryExecutionException =
+        Mockito.mock(AsyncApiException.class);
+      Mockito.when(mock.getQueryResultStream(jobId, batchId, resultIds[i]))
+        .thenThrow(salesforceQueryExecutionException);
+      Mockito.when(salesforceQueryExecutionException.getExceptionCode()).thenReturn(AsyncExceptionCode.Unknown);
+    }
+    reader.setupParser();
+  }
+
+  @Test
+  public void testSetupParserWithRetrySuccess() throws Exception {
+    MapToRecordTransformer transformer = new MapToRecordTransformer();
```

Comment:
The first usage of `transformer` seem to be at line 457, as a general practice, local variables shouldn't be initialized more than 2-3 lines before they are used.

Reply:
updated